packages/base-driver/lib/basedriver/helpers.js

```diff
@@ -66,73 +67,157 @@ async function retrieveHeaders (link) {
+  // more precise, but we don't need to be very precise here and also don't want to
+  // overuse RAM and have a performance drop.
+  if (isDir && await calculateFolderIntegrity(currentPath) >= expectedIntegrity?.folder) {
+    return true;
+  }
+  if (!isDir && await calculateFileIntegrity(currentPath) === expectedIntegrity?.file) {
+    return true;
+  }
+  return false;
+}
+
+/**
+ * @typedef {Object} ConfigureAppOptions
+ * @property {?Function} onPostProcess Optional function, which should be applied
+ * to the application after it is downloaded/preprocessed. This function may be async
+ * and is expected to accept single object parameter with the following properties:
+ *
+ * - cachedAppInfo: The information about the previously cached app instance (if exists):
+ *    - packageHash: SHA1 hash of the package if it is a file and not a folder
+ *    - lastModified: Optional Date instance, the value of file's `Last-Modified` header
+ *    - immutable: Optional boolean value. Contains true if the file has an `immutable` mark
+ *                 in `Cache-control` header
+ *    - maxAge: Optional integer representation of `maxAge` parameter in `Cache-control` header
+ *    - timestamp: The timestamp this item has been added to the cache (measured in Unix epoch
+ *                 milliseconds)
+ *    - integrity: An object containing either `file` property with SHA1 hash of the file
+ *                 or `folder` property with total amount of cached files and subfolders
+ *    - fullPath: the full path to the cached app
+ * - isUrl: Whether the app has been downloaded from a remote URL
+ * - headers: Optional headers object. Only present if `isUrl` is true and if the server
+ * responds to HEAD requests.
+ * - appPath: A string containing full path to the preprocessed application package (either
+ * downloaded or a local one)
+ *
+ * The function is expected to either return a falsy value, which means the app must not be
+ * cached and a fresh copy of it is downloaded each time. If this function returns an object
+ * containing `appPath` property then the integrity of it will be verified and stored into
+ * the cache.
+ * @property {!string[]} supportedExtensions List of supported application extensions (
+ * including starting dots)
+ */
+
+/**
+ * Prepares an app to be used in an automated test. The app gets cached automatically
+ * if it is an archive or if it is downloaded from an URL.
+ *
+ * @param {string} app Either a full path to the app or a remote URL
+ * @param {string|string[]|ConfigureAppOptions} options
+ * @returns The full path to the resulting application bundle
+ */
+async function configureApp (app, options = {}) {
   if (!_.isString(app)) {
     // immediately shortcircuit if not given an app
     return;
   }
-  if (!_.isArray(supportedAppExtensions)) {
-    supportedAppExtensions = [supportedAppExtensions];
+
+  let supportedAppExtensions;
+  const {
+    onPostProcess,
```

Comment:
post-processing of what?  maybe this should be something like `onAppConfigured`

Reply:
post processing of the app. Basically this stage is executed after we ensure the package is on the local FS or has been successfully downloaded

Reply:
OK.  it'd make sense then to use a more descriptive name anyway

Reply:
I cannot think about some name that would match better, maybe you'd have some proposals there. 
`onAppConfigured` does not describe the stage properly,  we'd need something more descriptive